src/spock_recovery.c

```
@@ -0,0 +1,489 @@
+ * never used for normal replication.
+ *
+ * If a recovery slot already exists for this database, returns success without
+ * creating a duplicate. This makes the function safe to call multiple times.
+ *
+ * Args:
+ *		database_name: Name of the database to create recovery slot for
+ *
+ * Returns:
+ *		true if slot exists (created now or previously)
+ *		false on error (logged as WARNING, not fatal)
+ */
+bool
+create_recovery_slot(const char *database_name)
+{
+	SpockRecoverySlotData *slot;
+	char	   *slot_name;
+	bool		success = false;
+	bool		cleanup_needed = false;
+
+	if (!SpockRecoveryCtx)
+	{
+		elog(WARNING, "recovery coordinator not initialized");
+		return false;
+	}
+
+	if (!database_name)
+	{
+		elog(WARNING, "database name cannot be NULL");
+		return false;
+	}
+
+	LWLockAcquire(SpockRecoveryCtx->lock, LW_EXCLUSIVE);
+
+	slot = &SpockRecoveryCtx->recovery_slot;
+
+	/* If slot already exists, we're done */
+	if (slot->active)
+	{
+		LWLockRelease(SpockRecoveryCtx->lock);
+		elog(LOG, "recovery slot already exists");
+		return true;
+	}
+
+	/* Initialize the slot metadata */
+	initialize_recovery_slot(slot, database_name);
+
+	/* Copy slot name for use outside the lock */
+	slot_name = pstrdup(slot->slot_name);
+	LWLockRelease(SpockRecoveryCtx->lock);
+
+	/*
+	 * Create the actual replication slot
+	 *
+	 * Parameters:
+	 * - slot_name: name of the slot to create
+	 * - true: database-specific slot
+	 * - RS_PERSISTENT: slot survives server restart
+	 * - false: not two-phase (we don't need prepared transaction support)
+	 * - false: not failover (recovery slots don't need failover)
+	 * - false: not synced (recovery slots are local only)
+	 */
+	PG_TRY();
+	{
+		ReplicationSlotCreate(slot_name, true, RS_PERSISTENT, false, false, false);
+		success = true;
+
+		elog(LOG, "created recovery slot '%s' (INACTIVE - for catastrophic failure recovery only)",
+			 slot_name);
+	}
+	PG_CATCH();
+	{
```

Let’s fail on error here. If recovery slots are enabled, failure to create one should be treated as a fatal error and result in an exit.

Using PG_TRY/CATCH just to swallow the failure complicates things unnecessarily; we can drop that and keep the logic simple.

Also, only touch shared-memory state after a successful create; that removes the success / cleanup_needed flags and the reset path entirely.